### PR TITLE
MueLu: Add matrix-free restriction and prolongation transfers

### PR DESCRIPTION
**Pull request by GrahamBenHarper**
@trilinos/muelu 

No longer WIP :)

## Motivation
This adds matrix-free grid transfers to MueLu. Officially, this adds two classes: `MatrixFreeTentativePFactory_kokkos` and `MatrixFreeTentativeP_kokkos`. These are two kokkos-based classes, one deriving from `PFactory` and one deriving from `Xpetra::Operator`. Unfortunately, this requires the Kokkos refactor as it depends on Kokkos views obtained from the `Aggregates_kokkos` object.

## Limitations
1. This is never to be used with aggregates that cross processors.
2. This currently only supports a single nullspace vector. There are some nontrivial changes that likely need to be made to do this for larger nullspaces as in the case of say, elasticity.
3. There is no way to generate a coarse operator from a matrix-free fine operator in MueLu yet, so these are strictly tested in similar cases where unsmoothed aggregation is applied.
4. Things such as strided maps, blocks, and others are not considered at the moment.

I'm planning to touch up a lot of these limitations as time goes on, but I'm hoping to get this merged this week so I can add a `MatrixFreeTentativeRAPFactory` so I can actually test the matrix-free hierarchy.

### CODE REVIEW

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativePFactory_kokkos_def.hpp
Comment (cgcgcg) [inline packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativePFactory_kokkos_def.hpp]: remove output, or keep, but use correct output stream

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (cgcgcg) [inline packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:131]: This should probably just throw cause it doesn't make much sense.

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (cgcgcg) [inline packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:102]: Do these loops have the correct nesting for memory access?

**Review by cgcgcg — CHANGES_REQUESTED** [review]

**Comment by cgcgcg** [comment]
Maybe wait until #10902 is merged?

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_decl.hpp
Comment (cgcgcg) [inline packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_decl.hpp:101]: I think you can use `= default` instead of an empty one.

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativePFactory_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativePFactory_kokkos_def.hpp]: Removed!

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:102]: Probably not. I think the right way to go is a `TeamPolicy` or an `MDRangePolicy`?

File: packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp
Comment (cgcgcg) [inline packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp:351]: Add a unit test that constructs both MatrixFree and matrix based operators and then times their applies?

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:131]: Yeah... it's an pure virtual function, so I have to override it, and it's a one-liner to do correctly, but I don't know if we use that interface anywhere in MueLu. This was my solution that does something correct despite not being needed.

**Comment by GrahamBenHarper** [comment]
> Maybe wait until #10902 is merged?

Yep. This is blocked by #10902

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (cgcgcg) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:131]: What's the residual supposed to be for a square matrix? I'd just throw.

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_decl.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_decl.hpp:101]: I think @mayrmt had a similar comment on #10833. I've been copying what is already in MueLu, and it seems like we tend to be more pedantic with writing constructors/destructors out, even if they're the default behavior. I don't know how consistently this is being done across MueLu, however.

File: packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp
Comment (GrahamBenHarper) [reply packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp:351]: Ooooo I like that suggestion. That would be good for answering the question above about access patterns too. I'll look into it.

Comment (cgcgcg) [reply packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp:351]: And while you're at it, you can check that they give the same result when applied to a random vector.

File: packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
Comment (cgcgcg) [inline packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp:119]: Are these includes needed? Cause it's not used here, is it?

File: packages/muelu/src/Interface/MueLu_ParameterListInterpreter_decl.hpp
Comment (cgcgcg) [inline packages/muelu/src/Interface/MueLu_ParameterListInterpreter_decl.hpp:112]: Same question

File: packages/muelu/src/MueCentral/MueLu_FactoryManager_decl.hpp
Comment (cgcgcg) [inline packages/muelu/src/MueCentral/MueLu_FactoryManager_decl.hpp:87]: Or here?

File: packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp:119]: Good question. I think it will need to be there eventually, but you're right, it's not officially a ParameterList option in this PR. I can remove it for now.

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (cgcgcg) [inline packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:92]: Couldn't you use `vertex2agg` here as well?

File: packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Interface/MueLu_ParameterListInterpreter_def.hpp]: Done!

File: packages/muelu/src/MueCentral/MueLu_FactoryManager_decl.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/MueCentral/MueLu_FactoryManager_decl.hpp]: Done!

File: packages/muelu/src/Interface/MueLu_ParameterListInterpreter_decl.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Interface/MueLu_ParameterListInterpreter_decl.hpp]: Done!

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:131]: oh oh good point. I strictly put the equation there without thinking about it. done!

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:92]: I made a comment about `sqrt`ing the whole view on line 101 actually :stuck_out_tongue: It's probably slightly more performant.

As for `vertex2agg`, I need to think about this. For some reason I'm confused. I could have sworn I needed `GetVertex2AggId` for one direction and `ComputeNodesInAggregate` for the other direction, and now I'm questioning everything

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp]: This should be done. I just went for an `MDRangePolicy` based on the column-major ordering of the multivector

File: packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp
Comment (GrahamBenHarper) [reply packages/muelu/test/unit_tests_kokkos/MatrixFree_kokkos.cpp]: This is done :)

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp:94]: I decided against doing a sqrt of the aggSize vector for now. I don't think it would be easy to justify an entire kernel launch for sqrting the aggSizes

**Comment by GrahamBenHarper** [comment]
Squashing 28 commits feels so nice :slightly_smiling_face: 
Now we'll see what the AT has to say...

File: packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_decl.hpp
Comment (GrahamBenHarper) [reply packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_decl.hpp]: Done!

**Review by cgcgcg — APPROVED** [review]
lgtm

**Review by cgcgcg — APPROVED** [review]
LGTM. Squash merge?

**Pull request merged**
